### DISCUSSION TITLE
What is the benchmark of query vector raw data by primary key?

### DISCUSSION THREAD

**Discussion by GitHuY** [milvus-io/milvus#24787] [Q&A and General discussion] [answered]
Our use case not only need to do a ANN search but also need to query the vector raw data by primary key. 
We find that the first time when query a key ("12345678", "223456") it will spend 1~2 seconds, but when we query another key("12345679", "223457") which is nearby the previous one, it only takes about 40 milli-seconds. The test is as below:

_python3 test-query.py 12345678
qeruy time(second) takes: 1.6750035285949707
 
python3 test-query.py 12345679
qeruy time(second) takes: 0.04695534706115723

python3 test-query.py 223456
qeruy time(second) takes: 1.308311939239502

python3 test-query.py 223457
qeruy time(second) takes: 0.04295921325683594_

1.  What's the benchmark of query vector raw data by primary key? 
2.  Do you have any suggestions for improving the latency of query by id?

**Comment by xiaofan-luan** [comment]
@GitHuY 
We don't have one right now, but it is a interesting topic to discuss.
How large is the dataset? Is there a way we can reproduce in house?
Or do you know what is the time consumption part?

**Reply by GitHuY** [reply]
@xiaofan-luan 
The dataset space is 30M records with 256 dimensions, the primary key range is from 1 to 30,000,000. We bulk insert all the records by the primary key order. You can try the queries like below :

for (int i = 0; i < 1000; i ++) {
       int pk = i * 20000 + i;
       query(pk);
       query(pk+1);
}

The 'query(pk)' is always slow, but the 'query(pk+1)' is fast. 

I'm not sure what is the most time consumption part, but I guess it's related to the segment load?

**Comment by xiaofan-luan** [comment]
did you require embeddings as the output field?

**Reply by GitHuY** [reply]
@xiaofan-luan
Yes, I require embeddings as output. I use Query API (not Search API) to get the embedding data by primary key.

**Reply by xiaofan-luan** [reply, marked as answer]
> @xiaofan-luan Yes, I require embeddings as output. I use Query API (not Search API) to get the embedding data by primary key.

You will see big improvement on Milvus 2.3 - We are gonna to release this month.
You will see the latency reduced to < 5ms for HNSW/IVF_FLAT and < 50ms for DiskANN

**Reply by GitHuY** [reply]
@xiaofan-luan 
Thanks. The latency (reduced to < 5ms for HNSW/IVF_FLAT) includes the embedding value as the output field?

**Reply by xiaofan-luan** [reply]
> @xiaofan-luan Thanks. The latency (reduced to < 5ms for HNSW/IVF_FLAT) includes the embedding value as the output field?

Yep, for hnsw it is doable because all the embedding is already in HNSW index

**Reply by GitHuY** [reply]
> > @xiaofan-luan Thanks. The latency (reduced to < 5ms for HNSW/IVF_FLAT) includes the embedding value as the output field?
> 
> Yep, for hnsw it is doable because all the embedding is already in HNSW index

I see, based on my understanding, it can be fast when do ANN search on this index. But how about only query the embedding by primary key (not do ANN search), which will not leverage on the HNSW/IVF index, is it also fast?

**Reply by xiaofan-luan** [reply]
> > > @xiaofan-luan Thanks. The latency (reduced to < 5ms for HNSW/IVF_FLAT) includes the embedding value as the output field?
> > 
> > 
> > Yep, for hnsw it is doable because all the embedding is already in HNSW index
> 
> I see, based on my understanding, it can be fast when do ANN search on this index. But how about only query the embedding by primary key (not do ANN search), which will not leverage on the HNSW/IVF index, is it also fast?

Maybe you could test on current master release or day build
